pkg/util/pki/csr.go

```diff
@@ -198,58 +219,95 @@ func GenerateCSR(crt *v1.Certificate, optFuncs ...GenerateCSROption) (*x509.Cert
 		opt(opts)
 	}
 
-	var (
-		commonName = crt.Spec.CommonName
-		err        error
-	)
-
-	if opts.UseLiteralSubject {
-		commonName, err = extractCommonNameFromLiteralSubject(crt.Spec)
+	// Generate the Subject field for the CSR.
+	var commonName string
+	var rdnSubject pkix.RDNSequence
+	if opts.UseLiteralSubject && len(crt.Spec.LiteralSubject) > 0 {
+		subjectRDNSequence, err := UnmarshalSubjectStringToRDNSequence(crt.Spec.LiteralSubject)
 		if err != nil {
 			return nil, err
 		}
 
-	}
+		commonName = ExtractCommonNameFromRDNSequence(subjectRDNSequence)
+		rdnSubject = subjectRDNSequence
+	} else {
+		subject := SubjectForCertificate(crt)
 
-	iPAddresses := IPAddressesForCertificate(crt)
-	organization := OrganizationForCertificate(crt)
-	subject := SubjectForCertificate(crt)
+		commonName = crt.Spec.CommonName
+		rdnSubject = pkix.Name{
+			Country:            subject.Countries,
+			Organization:       subject.Organizations,
+			OrganizationalUnit: subject.OrganizationalUnits,
+			Locality:           subject.Localities,
+			Province:           subject.Provinces,
+			StreetAddress:      subject.StreetAddresses,
+			PostalCode:         subject.PostalCodes,
+			SerialNumber:       subject.SerialNumber,
+			CommonName:         commonName,
+		}.ToRDNSequence()
+	}
 
-	dnsNames, err := DNSNamesForCertificate(crt)
+	// Generate the SANs for the CSR.
+	ipAddresses, err := IPAddressesFromStrings(crt.Spec.IPAddresses)
 	if err != nil {
 		return nil, err
 	}
 
-	uriNames, err := URIsForCertificate(crt)
+	uris, err := URLsFromStrings(crt.Spec.URIs)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(commonName) == 0 && len(dnsNames) == 0 && len(uriNames) == 0 && len(crt.Spec.EmailAddresses) == 0 && len(crt.Spec.IPAddresses) == 0 {
-		return nil, fmt.Errorf("no common name, DNS name, URI SAN, Email SAN or IP address specified on certificate")
+	if len(commonName) == 0 &&
+		len(crt.Spec.EmailAddresses) == 0 &&
+		len(crt.Spec.DNSNames) == 0 &&
+		len(uris) == 0 &&
+		len(ipAddresses) == 0 {
+		return nil, fmt.Errorf("no common name, DNS name, URI SAN, Email SAN, IP or OtherName SAN specified on certificate")
 	}
 
 	pubKeyAlgo, sigAlgo, err := SignatureAlgorithm(crt)
 	if err != nil {
 		return nil, err
 	}
 
+	asn1Subject, err := MarshalRDNSequenceToRawDERBytes(rdnSubject)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
suggestion: I'm not saying this is a huge blocker or anything, but the fact that we unconditionally set RawSubject here is a non-trivial change to the logic of this function. Previously we'd only set RawSubject if we were using a literal subject.

One approach to this refactor would be to copy paste the existing `GenerateCSR` function (call it `legacyGenerateCSR`, it doesn't have to be public) and then have a test to confirm that the output of `legacyGenerateCSR` matches the output of the new refactored `GenerateCSR`.

That would let us be confident that we're not changing the certificates we generate for people. Subject contents and even ordering of fields in subjects are things people care about and it would be a shame to break them.

Does that make sense? What do you think?

Reply:
Actually, I don't think that's needed. The underlying code just does what we're doing here: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/crypto/x509/x509.go;l=2038-2044

Never mind 😁